### ISSUE TITLE
Invalid argument on pull 

### ISSUE THREAD

**Issue by 1605200517** [moby/vpnkit#255]
Error log on a windows client:

```
[11:19:44.279][VpnKit         ][Error  ] vpnkit.exe: PPP.listen callback caught (Invalid_argument

[11:19:44.279][VpnKit         ][Info   ]   "Cstruct.blit src=[6296,1360](65536) dst=[0,4096](4096) dst-off=3673 `len=1360")
```

https://github.com/moby/moby/issues/34080

**Comment by djs55** [comment]
The version of vpnkit in 17.06.0 is befe8be9b45f6c4d50cf7c8105bfdf101dfe8982 which is built against [cstruct.2.4.1](https://github.com/mirage/ocaml-cstruct/tree/v2.4.1).

The specific error `"Cstruct.blit src=[6296,1360](65536) dst=[0,4096](4096) dst-off=3673 len=1360")` means
- copy `1360` bytes
- from offset `6296` in `src` which has a total length of `65536`
- to offset `3673` in `dst` which has a total length of `4096` and therefore only `423` bytes remaining. Since `423` is less than `1360` there is a problem.

The `65536` buffer is probably the [read_buffer](https://github.com/moby/vpnkit/blob/d660726b4ea1b7b20d044b773bd4723aa68e81fe/src/hostnet/host_uwt.ml#L362) created for the outgoing connection to the registry. The `4096` buffer is probably an "I/O page" destined for transmission to the VM.

The fact that this bug only manifests on Windows strongly suggests it is in the Windows-specific code, which is currently
- support for jumbo ethernet frames; and
- [Hyper-V socket bindings](https://github.com/mirage/ocaml-hvsock/tree/v0.13.0).

Regarding jumbo ethernet frame support, [this function](https://github.com/djs55/mirage-tcpip/blob/f5b16043c57f71d5fdf1a7cf7ab7ccdde291cca1/tcp/user_buffer.ml#L164) on the TCP transmit path has a suspicious `Cstruct.blit`. If a large frame is being assembled out of small parts, the whole thing is not guaranteed to fit in one 4096 byte page. I'll prepare an experimental build without jumbo frame support to see if this fixes the issue.

**Comment by djs55** [comment]
@1605200517 could you try an experiment? I'd like to see whether the problem is triggered by the large MTU setting on Windows.

With the most recent `17.06.0-ce-win19', download and decompress the attached  [vpnkit.zip](https://github.com/moby/vpnkit/files/1159848/vpnkit.zip). The SHA256 should be
```
9a263b4cacb3861530b4189d6bf0a5b6c824b97c2274912283595050a626deae  vpnkit.exe
```
(you can use `Get-FileHash vpnkit.exe` from powershell or `shasum -a 256` on OSX)

Then:
- stop Docker for Windows
- take a backup of the old file: `cp C:\Program Files\Docker\Docker\resources\vpnkit.exe c:\Program Files\Docker\Docker\resources\vpnkit.backup`
- replace the original file: `cp vpnkit.exe C:\Program Files\Docker\Docker\resources\vpnkit.exe`
- restart Docker for Windows

If everything starts ok, run a root shell with `docker run --rm --net=host --pid=host --privileged -it justincormack/nsenter1 /bin/sh` and type `ifconfig eth0 | grep MTU`. You should see a line like:
```
/ # ifconfig eth0 | grep MTU
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
```
-- note the MTU has been set to `1500`. Assuming that's ok, could you try pulling the image from the private registry which triggered the original problem? If it still fails, could you upload more diagnostics? Thanks!

**Comment by 1605200517** [comment]
hi, @djs55, all of the steps you have provided above produce excpected output.

The I did the test with today's release `17.06.1-ce-rc1-win20 (12818)' (edge channel) and it works!! thank you so much for this patch, you are awesome. :-) Works both for sequential and parallel download modes.

**Comment by djs55** [comment]
@1605200517 I'm glad everything worked for you! The patch should be in the future stable (17.06.1) and edge (17.07) updates. Let me know if anything else goes wrong.

Thanks for your report and quick feedback.

**Comment by 1605200517** [comment]
fixed also in Version 17.07.0-ce-rc1-win21 (12927)
Channel: edge
76a8c4b
